resolver/custom_dns_resolver.go

```diff
@@ -120,26 +163,108 @@ func (r *CustomDNSResolver) processRequest(request *model.Request) *model.Respon
+	ctx context.Context,
+	request *model.Request,
+	resolvedCnames []string,
+	question dns.Question,
+	entry dns.RR,
+) ([]dns.RR, error) {
+	switch v := entry.(type) {
+	case *dns.A:
+		return r.processIP(v.A, question)
+	case *dns.AAAA:
+		return r.processIP(v.AAAA, question)
+	case *dns.CNAME:
+		return r.processCNAME(ctx, request, *v, resolvedCnames, question)
+	}
+
+	return nil, fmt.Errorf("unsupported customDNS RR type %T", entry)
 }
 
 // Resolve uses internal mapping to resolve the query
 func (r *CustomDNSResolver) Resolve(ctx context.Context, request *model.Request) (*model.Response, error) {
-	logger := log.WithPrefix(request.Log, "custom_dns_resolver")
-
 	reverseResp := r.handleReverseDNS(request)
 	if reverseResp != nil {
 		return reverseResp, nil
 	}
 
-	if len(r.mapping) > 0 {
-		resp := r.processRequest(request)
-		if resp != nil {
-			return resp, nil
+	resp, err := r.processRequest(ctx, request, make([]string, 0, len(r.cfg.Mapping)))
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func (r *CustomDNSResolver) processIP(ip net.IP, question dns.Question) (result []dns.RR, err error) {
+	result = make([]dns.RR, 0)
+
+	if isSupportedType(ip, question) {
+		rr, err := r.createAnswerFromQuestion(question, ip, r.cfg.CustomTTL.SecondsU32())
+		if err != nil {
+			return nil, err
 		}
+
+		result = append(result, rr)
 	}
 
-	logger.WithField("next_resolver", Name(r.next)).Trace("go to next resolver")
+	return result, nil
+}
+
+func (r *CustomDNSResolver) processCNAME(
+	ctx context.Context,
+	request *model.Request,
+	targetCname dns.CNAME,
+	resolvedCnames []string,
+	question dns.Question,
+) (result []dns.RR, err error) {
+	cname := new(dns.CNAME)
+	ttl := r.cfg.CustomTTL.SecondsU32()
+	cname.Hdr = dns.RR_Header{Class: dns.ClassINET, Ttl: ttl, Rrtype: dns.TypeCNAME, Name: question.Name}
+	cname.Target = dns.Fqdn(targetCname.Target)
+	result = append(result, cname)
+
+	targetWithoutDot := strings.TrimSuffix(targetCname.Target, ".")
+
+	if slices.Contains(resolvedCnames, targetWithoutDot) {
+		return nil, fmt.Errorf("CNAME loop detected: %v", append(resolvedCnames, targetWithoutDot))
+	}
+
+	clientIP := request.ClientIP.String()
+	clientID := request.RequestClientID
+	aRequest := newRequestWithClientID(targetWithoutDot, dns.Type(dns.TypeA), clientIP, clientID)
+	aaaaRequest := newRequestWithClientID(targetWithoutDot, dns.Type(dns.TypeAAAA), clientIP, clientID)
+
+	cnames := resolvedCnames
+	cnames = append(cnames, targetWithoutDot)
+
+	if question.Qtype == dns.TypeA {
+		// Resolve target recursively
+		targetResp, err := r.processRequest(ctx, aRequest, cnames)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, targetResp.Res.Answer...)
+	}
+
+	if question.Qtype == dns.TypeAAAA {
+		// Resolve ipv6 target recursively
+		targetResp, err := r.processRequest(ctx, aaaaRequest, cnames)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, targetResp.Res.Answer...)
+	}
```

Comment:
I think we should do like the following to avoid duplication and be ready for other qtypes we add:
```suggestion
	cnames := resolvedCnames
	cnames = append(cnames, targetWithoutDot)

	// Resolve target recursively
	targetReq := newRequestWithClientID(targetWithoutDot, dns.Type(question.Qtype), clientIP, clientID)

	targetResp, err := r.processRequest(ctx, targetReq, cnames)
	if err != nil {
		return nil, err
	}

	result = append(result, targetResp.Res.Answer...)
```